src/WebJobs.Script/Host/HostIdValidator.cs

```diff
@@ -0,0 +1,198 @@
+                if (primaryHostStateProvider != null && !primaryHostStateProvider.IsPrimary)
+                {
+                    // as an optimization, only validate on the primary host
+                    return;
+                }
+
+                HostIdInfo hostIdInfo = await ReadHostIdInfoAsync(hostId);
+
+                if (hostIdInfo != null)
+                {
+                    // an existing record exists for this host ID
+                    // verify the host name is the same as our host name
+                    if (string.Compare(_hostNameProvider.Value, hostIdInfo.Hostname, StringComparison.OrdinalIgnoreCase) != 0)
+                    {
+                        HandleCollision(hostId);
+                    }
+                }
+                else
+                {
+                    // no existing record, so write one, claiming this host ID for this host name
+                    // in this storage account
+                    hostIdInfo = new HostIdInfo
+                    {
+                        Hostname = _hostNameProvider.Value
+                    };
+                    await WriteHostIdAsync(hostId, hostIdInfo);
+                }
+            }
+            catch (Exception ex)
+            {
+                // best effort - log error and continue
+                _logger.LogError(ex, "Error validating host ID usage.");
+            }
+        }
+
+        private void HandleCollision(string hostId)
+        {
+            // see if the user has specified a level, otherwise default
+            LogLevel level = DefaultLevel;
+            string value = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsHostIdCheckLevel);
+            if (value != null && Enum.TryParse<LogLevel>(value, out LogLevel parsedLevel))
+            {
+                level = parsedLevel;
+            }
+
+            string message = string.Format(Resources.HostIdCollisionFormat, hostId);
+            if (level == LogLevel.Error)
+            {
+                _logger.LogError(message);
+                _applicationLifetime.StopApplication();
+            }
+            else
+            {
+                // we only allow Warning/Error levels to be specified, so anything other than
+                // Error is treated as warning
+                _logger.LogWarning(message);
+            }
+        }
+
+        internal async Task WriteHostIdAsync(string hostId, HostIdInfo hostIdInfo)
+        {
+            try
+            {
+                var containerClient = _storageProvider.GetBlobContainerClient();
+                string blobPath = string.Format(BlobPathFormat, hostId);
+                BlobClient blobClient = containerClient.GetBlobClient(blobPath);
+                BinaryData data = BinaryData.FromObjectAsJson(hostIdInfo);
+                await blobClient.UploadAsync(data);
+            }
+            catch (Exception ex)
```

Comment:
After updating the overload to avoid overwriting the blob, we should handle the storage exception if the blob write fails and evaluate whether we need to shut this instance down (yes if the blob was written by another site, but we could repeat the check)

Reply:
Done